# PDF Link Extraction

- Testing a workflow to programmatically save all links share in PDf slides.
- Ideally want to save them to a Lecture/Week specific folder.

## Testing Approaches 

### Source: https://www.geeksforgeeks.org/extract-hyperlinks-from-pdf-in-python/
- Method 1: Using PyPDF2
- Method 2: Using pdfx

In [1]:
!pip install PyPDF2


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.0/80.0 kB 617.9 kB/s eta 0:00:00a 0:00:01


In [105]:
test_pdf = '../Week_01/Lecture_02/01.24.22 Week 1, Lecture 2_ Loading Data.pdf'
test_pdf

'../Week_01/Lecture_02/01.24.22 Week 1, Lecture 2_ Loading Data.pdf'

In [106]:
## Adapted from: https://www.geeksforgeeks.org/extract-hyperlinks-from-pdf-in-python/
import PyPDF2


file = test_pdf

pdfFileObject = open(file, 'rb')




In [107]:

pdfReader = PyPDF2.PdfFileReader(pdfFileObject)
help(pdfReader)

Help on PdfFileReader in module PyPDF2._reader object:

class PdfFileReader(builtins.object)
 |  PdfFileReader(stream, strict=True, warndest=None, overwriteWarnings=True)
 |  
 |  Initialize a PdfFileReader object.
 |  
 |  This operation can take some time, as the PDF stream's cross-reference
 |  tables are read into memory.
 |  
 |  :param stream: A File object or an object that supports the standard read
 |      and seek methods similar to a File object. Could also be a
 |      string representing a path to a PDF file.
 |  :param bool strict: Determines whether user should be warned of all
 |      problems and also causes some correctable problems to be fatal.
 |      Defaults to ``True``.
 |  :param warndest: Destination for logging warnings (defaults to
 |      ``sys.stderr``).
 |  :param bool overwriteWarnings: Determines whether to override Python's
 |      ``warnings.py`` module with a custom implementation (defaults to
 |      ``True``).
 |  
 |  Methods defined here:
 |  
 | 

In [108]:
pdfReader.numPages

30

In [111]:
page = pdfReader.getPage(-1)
page.extractText()

'\n\n\x02¿\n\n\n\n\n\n\n\n'

In [112]:
# https://github.com/py-pdf/PyPDF2/issues/437
!pip install slate3k


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.8/100.8 kB 1.9 MB/s eta 0:00:00a 0:00:01


In [115]:

import slate3k as slate
slate.PDF?

Init signature:
slate.PDF(
    file,
    password='',
    just_text=1,
    check_extractable=True,
    char_margin=1.0,
    line_margin=0.1,
    word_margin=0.1,
)
Docstring:     
Built-in mutable sequence.

If no argument is given, the constructor creates a new empty list.
The argument must be an iterable if specified.
File:           /opt/homebrew/Caskroom/miniforge/base/envs/dev-env/lib/python3.8/site-packages/slate3k/classes.py
Type:           type
Subclasses:     


In [116]:

# import slate3k as slate

# with open(test_pdf,'rb') as f:
#     extracted_text = slate.PDF(f )
# print(extracted_text)

In [12]:
print(dir(page))

['__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_addTransformationMatrix', '_contentStreamRename', '_mergePage', '_mergeResources', '_pushPopGS', '_rotate', 'addTransformation', 'artBox', 'bleedBox', 'clear', 'compressContentStreams', 'copy', 'createBlankPage', 'cropBox', 'extractText', 'fromkeys', 'get', 'getContents', 'getObject', 'getXmpMetadata', 'indirectRef', 'items', 'keys', 'mediaBox', 'mergePage', 'mergeRotatedPage', 'mergeRotatedScaledPage', 'mergeRotatedScaledTranslatedPage', 'mergeRotatedTranslatedPage', 'mergeScaledPage', 'mergeScaledTranslatedPage', 'mergeTransformedPage', 'mergeTr

In [13]:
page.extractText()

'\n\n\x01fl\n\x01£\x01£\x01G\x019\x01j\x01h\x01˚\x01j\x01±\x01£\x03x\n\n\x01fl\n\x01˚\x01h\x01˚\x01j\x01±\x01£\n\n\n\n\x01fl\n\x01Ø\x03I\n\n\x01|\n\x01˚\x01£\x01£\x01G\x01u\x01j\x03N\n\n\x01|\n\x01˚\x01£\x01£\x01G\x01u\x01j\x03x\x01Ò\x01u\x01ł\x01[\x03N\n\n\x01|\n\x01˚\x03x\x01\x10\x01u\x01˚\x018\x01è\x01\x10\x01G\x01˚\x01j\x01±\x01£\x03N\n\n\x01|\n\x01˚\x01£\x01£\x01G\x01u\x01j\n\n\x01|\n\x01Ò\x03x\x01±\x01˚\x019\x01ł\n\n\n\x01fl\n\n\n\x01|\n\x01˚\x01£\x01£\x01G\x01u\x01j\x03x\x01h\x01u\x01\x17\x01˚\x01_\n\n\x01|\n\x01˚\x01ł\n\n\x01|\n\x01\x17\x01Ñ\x01u\x01A\x01˚\x01£\n\n\n'

In [14]:
page.getContents()

{'/Filter': '/FlateDecode'}

In [ ]:

## save results in dictionary
text = {}

for page_number in range(pdfReader.numPages):
	
	pageObject = pdfReader.getPage(page_number)
	pdf_text = pageObject.extractText()
	text[page_number] = pdf_text
	
pdfFileObject.close()


In [4]:
pdf_text

'\n\n\n\n\n\n\n\n\x01fl\n\n\n\x01|\n\n\x01|\n\n\x01|\n\n\n\n\x01fl\n\n\x01|\n\n\x01|\n\n\x01|\n\n\n'

## Method 2

> Note: process works, but doesn't get links that have different display text

In [15]:
!pip install pdfx


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 25.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.3/5.3 MB 34.2 MB/s eta 0:00:0000:0100:01
  Using cached sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)


In [16]:
# Import Module
import pdfx

# Read PDF File
pdf = pdfx.PDFx(test_pdf)

# Get list of URL
print(pdf.get_references_as_dict())


{'url': ['https://github.com/coding-dojo-data-science/data-enrichment-linear-regression-with-movies']}


In [38]:
import glob,os,re

all_pdfs = sorted(glob.glob('../Week*/**/*.pdf',recursive=True))
all_pdfs

['../Week_01/Lecture_01/01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf',
 '../Week_01/Lecture_02/01.24.22 Week 1, Lecture 2_ Loading Data.pdf',
 '../Week_02/Lecture_01/01.24.22 Week 2, Day 1_ Pandas .pdf',
 '../Week_02/Lecture_02/_01.24.22 Week 2, Day 2_ Data Cleaning  - Final.pdf',
 '../Week_03/Lecture_01/01.24.22 Week 3 Day 1_ Clean and Univariate Exploration (1).pdf',
 '../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf',
 '../Week_04/Lecture_01/Week 4 Day 1_ Viz Best Practices.pdf',
 '../Week_05/Lecture_01/Week 5 Day 1_ ML and Model Validation.pdf',
 '../Week_05/Lecture_02/1.24 Week 5 Day 2_ Categorical Variables and Intro to ML.pdf',
 '../Week_06/Lecture_01/Week 6 Day 1_ Regression Metrics.pdf',
 '../Week_06/Lecture_02/Week 6 Day 2_  Bias_Variance and Regression  Practice.pdf',
 '../Week_07/Lecture_01/Week 7 Day 1_ Classification.pdf',
 '../Week_07/Lecture_02/PT DS Week  7 Day 2_ GridSearchCV and Pipelines.pdf',
 '../Week_08/Lecture_01/Week 8 Day 1

In [39]:
##test file
filename=all_pdfs[0]
filename

'../Week_01/Lecture_01/01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf'

In [40]:
os.path.basename(filename)

'01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf'

In [51]:
folder = os.path.dirname(filename)
week = re.findall(r"Week_\d{2}",folder)
lect = re.findall(r"Lect\w*_\d{2}",folder)
lect

['Lecture_01']

In [103]:
##test file
filename=all_pdfs[1]
filename


'../Week_01/Lecture_02/01.24.22 Week 1, Lecture 2_ Loading Data.pdf'

In [ ]:

pdf = pdfx.PDFx(filename)
links = pdf.get_references_as_dict()
print(pdf.get_text())

In [102]:
pdf.get_references_as_dict()

{'url': ['https://github.com/sensei-jirving/code-reviews-week-1-python-basics',
  'df.info']}

In [70]:
data = [['Week','Lecture','Title','Filepath','Links']]
for filename in all_pdfs:
	
	folder = os.path.dirname(filename)
	title = os.path.basename(filename)
	week = re.findall(r"Week_\d{2}",folder)[0]
	lect = re.findall(r"Lect\w*_\d{2}",folder)[0]
	
	pdf = pdfx.PDFx(filename)
	links = pdf.get_references_as_dict()
	
	if 'url' in links:
		link_list = links['url']
		link_list = [l for l in link_list if 'df.info' not in l]
	else:
		link_list = ['']
	data.append([week, lect, title, filename,link_list])
	

In [81]:
import pandas as pd
raw_df = pd.DataFrame(data[1:], columns=data[0])
raw_df

Week     Lecture                                              Title  \
0   Week_01  Lecture_01  01.24.22 Week 1, Lecture 1 - Python for Data S...   
1   Week_01  Lecture_02       01.24.22 Week 1, Lecture 2_ Loading Data.pdf   
2   Week_02  Lecture_01                01.24.22 Week 2, Day 1_ Pandas .pdf   
3   Week_02  Lecture_02  _01.24.22 Week 2, Day 2_ Data Cleaning  - Fina...   
4   Week_03  Lecture_01  01.24.22 Week 3 Day 1_ Clean and Univariate Ex...   
5   Week_03  Lecture_02  01.24.22 Week 3 Day 2_ Multivariate Exploratio...   
6   Week_04  Lecture_01               Week 4 Day 1_ Viz Best Practices.pdf   
7   Week_05  Lecture_01          Week 5 Day 1_ ML and Model Validation.pdf   
8   Week_05  Lecture_02  1.24 Week 5 Day 2_ Categorical Variables and I...   
9   Week_06  Lecture_01               Week 6 Day 1_ Regression Metrics.pdf   
10  Week_06  Lecture_02  Week 6 Day 2_  Bias_Variance and Regression  P...   
11  Week_07  Lecture_01                   Week 7 Day 1_ Classification.pdf   
12  Week_07  Lecture_02  PT DS Week  7 Day 2_ GridSearchCV and Pipeline...   
13  Week_08  Lecture_01      Week 8 Day 1_ Logistic Regression and ROC.pdf   
14  Week_08  Lecture_02                 Week 8 Day 2 Gradient Boosting.pdf   
15  Week_09  Lecture_01  01.24 Week 9 Day 1 _  Data Analysis with Clust...   
16  Week_09  Lecture_02  01.24 PT DS Week 9 Day 2_ Clustering for Super...   
17  Week_13  Lecture_01                        Wk13-L01 Intro to MySQL.pdf   
18  Week_13  Lecture_02                       Wk-13-L02 Database Admin.pdf   
19  Week_14  Lecture_01                        Wk 14-L01 - JSON & APIS.pdf   
20  Week_15  Lecture_01                     Wk 15-L01 - Mock Belt Exam.pdf   
21  Week_15  Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
22  Week_16  Lecture_01      Wk 16-L01 - Linear Regression - Revisited.pdf   
23  Week_16  Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   

                                             Filepath  \
0   ../Week_01/Lecture_01/01.24.22 Week 1, Lecture...   
1   ../Week_01/Lecture_02/01.24.22 Week 1, Lecture...   
2   ../Week_02/Lecture_01/01.24.22 Week 2, Day 1_ ...   
3   ../Week_02/Lecture_02/_01.24.22 Week 2, Day 2_...   
4   ../Week_03/Lecture_01/01.24.22 Week 3 Day 1_ C...   
5   ../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ M...   
6   ../Week_04/Lecture_01/Week 4 Day 1_ Viz Best P...   
7   ../Week_05/Lecture_01/Week 5 Day 1_ ML and Mod...   
8   ../Week_05/Lecture_02/1.24 Week 5 Day 2_ Categ...   
9   ../Week_06/Lecture_01/Week 6 Day 1_ Regression...   
10  ../Week_06/Lecture_02/Week 6 Day 2_  Bias_Vari...   
11  ../Week_07/Lecture_01/Week 7 Day 1_ Classifica...   
12  ../Week_07/Lecture_02/PT DS Week  7 Day 2_ Gri...   
13  ../Week_08/Lecture_01/Week 8 Day 1_ Logistic R...   
14  ../Week_08/Lecture_02/Week 8 Day 2 Gradient Bo...   
15  ../Week_09/Lecture_01/01.24 Week 9 Day 1 _  Da...   
16  ../Week_09/Lecture_02/01.24 PT DS Week 9 Day 2...   
17  ../Week_13/Lecture_01/Wk13-L01 Intro to MySQL.pdf   
18  ../Week_13/Lecture_02/Wk-13-L02 Database Admin...   
19  ../Week_14/Lecture_01/Wk 14-L01 - JSON & APIS.pdf   
20  ../Week_15/Lecture_01/Wk 15-L01 - Mock Belt Ex...   
21  ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis T...   
22  ../Week_16/Lecture_01/Wk 16-L01 - Linear Regre...   
23  ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing a...   

                                                Links  
0   [https://login.codingdojo.com/, https://bit.ly...  
1   [https://github.com/sensei-jirving/code-review...  
2                                                  []  
3   [https://bookdown.org/ejvanholm/Textbook/displ...  
4                                                  []  
5   [https://github.com/sensei-jirving/Online-DS-P...  
6   [https://www.python-graph-gallery.com/, https:...  
7                                                  []  
8                                                  []  
9                                                  []  
10            

In [82]:
df  =raw_df.explode('Links')
df = df.set_index(['Week','Lecture'])
df

Title  \
Week    Lecture                                                         
Week_01 Lecture_01  01.24.22 Week 1, Lecture 1 - Python for Data S...   
        Lecture_01  01.24.22 Week 1, Lecture 1 - Python for Data S...   
        Lecture_01  01.24.22 Week 1, Lecture 1 - Python for Data S...   
        Lecture_02       01.24.22 Week 1, Lecture 2_ Loading Data.pdf   
Week_02 Lecture_01                01.24.22 Week 2, Day 1_ Pandas .pdf   
        Lecture_02  _01.24.22 Week 2, Day 2_ Data Cleaning  - Fina...   
Week_03 Lecture_01  01.24.22 Week 3 Day 1_ Clean and Univariate Ex...   
        Lecture_02  01.24.22 Week 3 Day 2_ Multivariate Exploratio...   
        Lecture_02  01.24.22 Week 3 Day 2_ Multivariate Exploratio...   
        Lecture_02  01.24.22 Week 3 Day 2_ Multivariate Exploratio...   
        Lecture_02  01.24.22 Week 3 Day 2_ Multivariate Exploratio...   
Week_04 Lecture_01               Week 4 Day 1_ Viz Best Practices.pdf   
        Lecture_01               Week 4 Day 1_ Viz Best Practices.pdf   
        Lecture_01               Week 4 Day 1_ Viz Best Practices.pdf   
        Lecture_01               Week 4 Day 1_ Viz Best Practices.pdf   
Week_05 Lecture_01          Week 5 Day 1_ ML and Model Validation.pdf   
        Lecture_02  1.24 Week 5 Day 2_ Categorical Variables and I...   
Week_06 Lecture_01               Week 6 Day 1_ Regression Metrics.pdf   
        Lecture_02  Week 6 Day 2_  Bias_Variance and Regression  P...   
Week_07 Lecture_01                   Week 7 Day 1_ Classification.pdf   
        Lecture_02  PT DS Week  7 Day 2_ GridSearchCV and Pipeline...   
Week_08 Lecture_01      Week 8 Day 1_ Logistic Regression and ROC.pdf   
        Lecture_02                 Week 8 Day 2 Gradient Boosting.pdf   
Week_09 Lecture_01  01.24 Week 9 Day 1 _  Data Analysis with Clust...   
        Lecture_02  01.24 PT DS Week 9 Day 2_ Clustering for Super...   
Week_13 Lecture_01                        Wk13-L01 Intro to MySQL.pdf   
        Lecture_01                        Wk13-L01 Intro to MySQL.pdf   
        Lecture_02                       Wk-13-L02 Database Admin.pdf   
        Lecture_02                       Wk-13-L02 Database Admin.pdf   
        Lecture_02                       Wk-13-L02 Database Admin.pdf   
Week_14 Lecture_01                        Wk 14-L01 - JSON & APIS.pdf   
        Lecture_01                        Wk 14-L01 - JSON & APIS.pdf   
Week_15 Lecture_01                     Wk 15-L01 - Mock Belt Exam.pdf   
        Lecture_01                     Wk 15-L01 - Mock Belt Exam.pdf   
        Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
        Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
        Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
        Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
        Lecture_02     Wk 15-L02 - Hypothesis Testing with Python.pdf   
Week_16 Lecture_01      Wk 16-L01 - Linear Regression - Revisited.pdf   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   
        Lecture_02  Wk 16-L02 - Diagnosing and Interpreting Linear...   

                                                             Filepath  \
Week    Lecture                                                         
Week_01 Lecture_01  ../Week_01/Lecture_01/01.24.22 Week 1, Lecture...   
        Lecture_01  ../Week_01/Lecture_01/01.24.22 Week 1, Lecture...   
        Lecture_01  ../Week_01/Lecture_01/01.24.22 Week 1, Lecture...   
        Lecture_02  ../Week_01/Lecture_02/01.24.22 Week 1, Lecture...   
Week_02 Lecture_01  ../Week_02/Lecture_01/01.24.22 Week 2, Day 1_ ...   
        Lecture_02  ../Week_02/Lecture_02/_01.24

In [87]:
# !pip install tabulate
html = df.to_html()
with open('Links from Lecture Slides.html','w') as f:
	f.write(html)

In [88]:
df.to_csv('Links from Lecture Slides.csv')

In [91]:
txt = df.reset_index().drop(columns=["Filepath"]).to_markdown()
with open('Links from Lecture Slides.md','w') as f:
	f.write(txt)

|                                                                                                    | Filepath                                                                                    | Links                                                                                            |
|:---------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------|
| ('Week_01', 'Lecture_01', '01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf')              | ../Week_01/Lecture_01/01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf              | https://login.codingdojo.com/                                                                    |
| ('Week_01', 'Lecture_01', '01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf')              | ../Week_01/Lecture_01/01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf              | https://bit.ly/32k7fwU                                                                           |
| ('Week_01', 'Lecture_01', '01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf')              | ../Week_01/Lecture_01/01.24.22 Week 1, Lecture 1 - Python for Data Science.pdf              | https://commons.wikimedia.org/wiki/File:Six-handed_500_playing_cards.jpg                         |
| ('Week_01', 'Lecture_02', '01.24.22 Week 1, Lecture 2_ Loading Data.pdf')                          | ../Week_01/Lecture_02/01.24.22 Week 1, Lecture 2_ Loading Data.pdf                          | https://github.com/sensei-jirving/code-reviews-week-1-python-basics                              |
| ('Week_02', 'Lecture_01', '01.24.22 Week 2, Day 1_ Pandas .pdf')                                   | ../Week_02/Lecture_01/01.24.22 Week 2, Day 1_ Pandas .pdf                                   | nan                                                                                              |
| ('Week_02', 'Lecture_02', '_01.24.22 Week 2, Day 2_ Data Cleaning  - Final.pdf')                   | ../Week_02/Lecture_02/_01.24.22 Week 2, Day 2_ Data Cleaning  - Final.pdf                   | https://bookdown.org/ejvanholm/Textbook/displaying-data.html                                     |
| ('Week_03', 'Lecture_01', '01.24.22 Week 3 Day 1_ Clean and Univariate Exploration (1).pdf')       | ../Week_03/Lecture_01/01.24.22 Week 3 Day 1_ Clean and Univariate Exploration (1).pdf       |                                                                                                  |
| ('Week_03', 'Lecture_02', '01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf')                   | ../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf                   | https://github.com/sensei-jirving/Online-DS-PT-01.24.22-cohort-notes                             |
| ('Week_03', 'Lecture_02', '01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf')                   | ../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf                   | https://stackoverflow.com/questions/19966018/pandas-filling-missing-values-by-mean-in-each-group |
| ('Week_03', 'Lecture_02', '01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf')                   | ../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf                   | https://mstranslate.com.au/correlation-vs-causation/                                             |
| ('Week_03', 'Lecture_02', '01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf')                   | ../Week_03/Lecture_02/01.24.22 Week 3 Day 2_ Multivariate Exploration.pdf                   | https://login.codingdojo.com/exams                                                               |
| ('Week_04', 'Lecture_01', 'Week 4 Day 1_ Viz Best Practices.pdf')                                  | ../Week_04/Lecture_01/Week 4 Day 1_ Viz Best Practices.pdf                                  | https://www.python-graph-gallery.com/                                                            |
| ('Week_04', 'Lecture_01', 'Week 4 Day 1_ Viz Best Practices.pdf')                                  | ../Week_04/Lecture_01/Week 4 Day 1_ Viz Best Practices.pdf                                  | https://seaborn.pydata.org/examples/index.html                                                   |
| ('Week_04', 'Lecture_01', 'Week 4 Day 1_ Viz Best Practices.pdf')                                  | ../Week_04/Lecture_01/Week 4 Day 1_ Viz Best Practices.pdf                                  | https://matplotlib.org/stable/gallery/index.html                                                 |
| ('Week_04', 'Lecture_01', 'Week 4 Day 1_ Viz Best Practices.pdf')                                  | ../Week_04/Lecture_01/Week 4 Day 1_ Viz Best Practices.pdf                                  | https://datavizproject.com/                                                                      |
| ('Week_05', 'Lecture_01', 'Week 5 Day 1_ ML and Model Validation.pdf')                             | ../Week_05/Lecture_01/Week 5 Day 1_ ML and Model Validation.pdf                             |                                                                                                  |
| ('Week_05', 'Lecture_02', '1.24 Week 5 Day 2_ Categorical Variables and Intro to ML.pdf')          | ../Week_05/Lecture_02/1.24 Week 5 Day 2_ Categorical Variables and Intro to ML.pdf          |                                                                                                  |
| ('Week_06', 'Lecture_01', 'Week 6 Day 1_ Regression Metrics.pdf')                                  | ../Week_06/Lecture_01/Week 6 Day 1_ Regression Metrics.pdf                                  |                                                                                                  |
| ('Week_06', 'Lecture_02', 'Week 6 Day 2_  Bias_Variance and Regression  Practice.pdf')             | ../Week_06/Lecture_02/Week 6 Day 2_  Bias_Variance and Regression  Practice.pdf             |                                                                                                  |
| ('Week_07', 'Lecture_01', 'Week 7 Day 1_ Classification.pdf')                                      | ../Week_07/Lecture_01/Week 7 Day 1_ Classification.pdf                                      |                                                                                                  |
| ('Week_07', 'Lecture_02', 'PT DS Week  7 Day 2_ GridSearchCV and Pipelines.pdf')                   | ../Week_07/Lecture_02/PT DS Week  7 Day 2_ GridSearchCV and Pipelines.pdf                   |                                                                                                  |
| ('Week_08', 'Lecture_01', 'Week 8 Day 1_ Logistic Regression and ROC.pdf')                         | ../Week_08/Lecture_01/Week 8 Day 1_ Logistic Regression and ROC.pdf                         |                                                                                                  |
| ('Week_08', 'Lecture_02', 'Week 8 Day 2 Gradient Boosting.pdf')                                    | ../Week_08/Lecture_02/Week 8 Day 2 Gradient Boosting.pdf                                    |                                                                                                  |
| ('Week_09', 'Lecture_01', '01.24 Week 9 Day 1 _  Data Analysis with Clustering.pptx.pdf')          | ../Week_09/Lecture_01/01.24 Week 9 Day 1 _  Data Analysis with Clustering.pptx.pdf          |                                                                                                  |
| ('Week_09', 'Lecture_02', '01.24 PT DS Week 9 Day 2_ Clustering for Supervised Learning.pptx.pdf') | ../Week_09/Lecture_02/01.24 PT DS Week 9 Day 2_ Clustering for Supervised Learning.pptx.pdf |                                                                                                  |
| ('Week_13', 'Lecture_01', 'Wk13-L01 Intro to MySQL.pdf')                                           | ../Week_13/Lecture_01/Wk13-L01 Intro to MySQL.pdf                                           | GitHub.com                                                                                       |
| ('Week_13', 'Lecture_01', 'Wk13-L01 Intro to MySQL.pdf')                                           | ../Week_13/Lecture_01/Wk13-L01 Intro to MySQL.pdf                                           | https://github.com/coding-dojo-data-science/data-enrichment-wk13-l01-activity                    |
| ('Week_13', 'Lecture_02', 'Wk-13-L02 Database Admin.pdf')                                          | ../Week_13/Lecture_02/Wk-13-L02 Database Admin.pdf                                          | databasestar.com                                                                                 |
| ('Week_13', 'Lecture_02', 'Wk-13-L02 Database Admin.pdf')                                          | ../Week_13/Lecture_02/Wk-13-L02 Database Admin.pdf                                          | https://sqlzoo.net/wiki/SQL_Tutorial                                                             |
| ('Week_13', 'Lecture_02', 'Wk-13-L02 Database Admin.pdf')                                          | ../Week_13/Lecture_02/Wk-13-L02 Database Admin.pdf                                          | c.name                                                                                           |
| ('Week_14', 'Lecture_01', 'Wk 14-L01 - JSON & APIS.pdf')                                           | ../Week_14/Lecture_01/Wk 14-L01 - JSON & APIS.pdf                                           | https://github.com/coding-dojo-data-science/data-enrichment-wk14-l01-json-activity               |
| ('Week_14', 'Lecture_01', 'Wk 14-L01 - JSON & APIS.pdf')                                           | ../Week_14/Lecture_01/Wk 14-L01 - JSON & APIS.pdf                                           | https://github.com/coding-dojo-data-science/data-enrichment-wk14-mapping-yelp-results            |
| ('Week_15', 'Lecture_01', 'Wk 15-L01 - Mock Belt Exam.pdf')                                        | ../Week_15/Lecture_01/Wk 15-L01 - Mock Belt Exam.pdf                                        | https://github.com/coding-dojo-data-science/data-enrichment-mock-belt-exam                       |
| ('Week_15', 'Lecture_01', 'Wk 15-L01 - Mock Belt Exam.pdf')                                        | ../Week_15/Lecture_01/Wk 15-L01 - Mock Belt Exam.pdf                                        | github.com                                                                                       |
| ('Week_15', 'Lecture_02', 'Wk 15-L02 - Hypothesis Testing with Python.pdf')                        | ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis Testing with Python.pdf                        | https://login.codingdojo.com/m/376/12528/88194                                                   |
| ('Week_15', 'Lecture_02', 'Wk 15-L02 - Hypothesis Testing with Python.pdf')                        | ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis Testing with Python.pdf                        | https://github.com/coding-dojo-data-science/data-enrichment-mock-belt-exam                       |
| ('Week_15', 'Lecture_02', 'Wk 15-L02 - Hypothesis Testing with Python.pdf')                        | ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis Testing with Python.pdf                        | https://forms.gle/bUxX4rznSEkpgaPR9                                                              |
| ('Week_15', 'Lecture_02', 'Wk 15-L02 - Hypothesis Testing with Python.pdf')                        | ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis Testing with Python.pdf                        | https://github.com/coding-dojo-data-science/data-enrichment-hypothesis-testing-codealong         |
| ('Week_15', 'Lecture_02', 'Wk 15-L02 - Hypothesis Testing with Python.pdf')                        | ../Week_15/Lecture_02/Wk 15-L02 - Hypothesis Testing with Python.pdf                        | github.com                                                                                       |
| ('Week_16', 'Lecture_01', 'Wk 16-L01 - Linear Regression - Revisited.pdf')                         | ../Week_16/Lecture_01/Wk 16-L01 - Linear Regression - Revisited.pdf                         | https://github.com/coding-dojo-data-science/data-enrichment-linear-regression-with-movies        |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://github.com/sensei-jirving/Online-DS-PT-01.24.22-cohort-notes                             |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://github.com/coding-dojo-data-science/data-enrichment-linear-regression-with-movies        |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://youtube.com/playlist?list=PLmeeqPbYmMC0XlmuN4agv0zvuAXP8HZS_                             |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://www.geeksforgeeks.org/detecting-multicollinear                                           |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://login.codingdojo.com/dashboard/end-stack-survey                                          |
| ('Week_16', 'Lecture_02', 'Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf')         | ../Week_16/Lecture_02/Wk 16-L02 - Diagnosing and Interpreting Linear Regression.pdf         | https://towardsdatascience.co                                                                    |
